trappy/cpu_power.py

```diff
@@ -120,6 +143,98 @@ def get_all_freqs(self, mapping_label):
 
         return pivot_with_labels(dfr, "freq", "cpus", mapping_label) / 1000
 
+    def plot_cdev_states(self, title="", width=None, height=None, xlim="default",
+                         ylim="range", legend_label="", drawstyle="default",
+                         cpus=None):
+        """Plot the cooling device state evolution
+
+        :param ax: Axis instance
+        :type ax: :mod:`matplotlib.Axis`
+
+        :param legend_label: Label for the legend
+        :type legend_label: str
+
+        :param title: The title of the plot
+        :type title: str
+
+        :param width: The width of the plot
+        :type width: int
+
+        :param height: The height of the plot
+        :type height: int
+
+        :param xlim: The xlim setting of the plot.
+            See :func:`~trappy.plot_utils.set_lim`
+        :type xlim: str or tuple of int
+
+        :param ylim: The ylim setting of the plot
+            See :func:`~trappy.plot_utils.set_lim`
+        :type ylim: str or tuple of int
+
+        :param drawstyle: The drawstyle setting of the plot
+        :type drawstyle: str
+
+        :param cpus: List of cpus to plot
+            All are plotted by default
+        :type cpus: list of int
+        """
+        from matplotlib import pyplot as plt
+        from trappy.plot_utils import normalize_title, pre_plot_setup, post_plot_setup
+
+        if len(self.data_frame) == 0:
+            raise ValueError("Empty DataFrame")
+
+        thermal_dfr = self.data_frame.copy()
+
+        # Sanitize cpumasks
+        thermal_dfr["cpus"] = thermal_dfr["cpus"].apply(sanitize_mask)
+
+        # Find available cpumasks
+        available_masks = []
+        for mask in thermal_dfr["cpus"].unique().tolist():
+            available_masks.append(mask)
+
+        # Sanitize cpus
+        if cpus is not None:
+            global_mask = cpus_to_mask(cpus)
+
+            # Find masks that match the requested CPUs
+            # This can include other CPUs
+            selected_masks = [m for m in available_masks if m & global_mask]
+
+            if len(selected_masks) == 0:
+                raise ValueError("No {} trace for CPUs {}".format(self.unique_word, cpus))
+
+            thermal_dfr = thermal_dfr[
+                thermal_dfr["cpus"].isin(selected_masks)
+            ]
+
+        cdevs = {}
+        # At this point we've removed the cpumasks we don't care about
+        for mask in thermal_dfr["cpus"].unique().tolist():
+            cdevs[mask] = thermal_dfr[thermal_dfr["cpus"] == mask]["cdev_state"]
+
+        # Prepare one subplot per cdev
+        ax = pre_plot_setup(width, height, nrows = len(cdevs))
```

Comment:
There's one thing I'm still a bit unsure about - the management of the axes. The other plotting methods I used as a starting point accept an `ax` parameter, and use that handle to plot stuff if it's `not None`. Thing is, I feel obligated to done one plot per cooling device because it gets unreadable otherwise (have look at the bottom of https://gist.github.com/valschneider/9f93ada029c8662ea954178906c4928f to see what I mean).

So that means that this `ax` parameter should be an array of axes if there's more than one cdev, so the user is expected to look into the dataframe data - and same goes for legends and titles, which is left to the user in the pre-existing methods. I went with a more "ignorant user" approach, but it may clash with the logic that has been followed for the other plots.

Reply:
Hmm, I wouldn't worry about it too much, it's definitely nice to be able to pass in your own axes but it's probably a niche requirement here. 

If you really wanted to do it right, I have a feeling that where you need multiple _axes_, what you need is a _figure_. But I don't understand matplotlib very well so that might be wrong, and if it's right I don't know how you'd actually use it.